micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaConsumerMetrics.java

```diff
@@ -0,0 +1,182 @@
+ */
+package io.micrometer.core.instrument.binder.kafka;
+
+
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import io.micrometer.core.lang.NonNullApi;
+import io.micrometer.core.lang.NonNullFields;
+
+import javax.management.*;
+import java.lang.management.ManagementFactory;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+
+import static java.util.Collections.emptyList;
+/**
+ * Kafka Consumer metrics collected from metrics exposed by org.apache.kafka.clients.consumer.KafkaConsumer
+ * via the MBeanServer. Metrics are exposed at each consumer thread.
+ *
+ * @author Wardha Perinkadakattu
+ */
+
+@NonNullApi
+@NonNullFields
+public class KafkaConsumerMetrics implements MeterBinder {
+
+    private final MBeanServer mBeanServer;
+
+    private final Iterable<Tag> tags;
+
+    public KafkaConsumerMetrics() {
+        this(getMBeanServer(), emptyList());
+    }
+
+    public KafkaConsumerMetrics(Iterable<Tag> tags) {
+        this(getMBeanServer(), tags);
+    }
+
+    public KafkaConsumerMetrics(MBeanServer mBeanServer, Iterable<Tag> tags) {
+        this.tags = tags;
+        this.mBeanServer = mBeanServer;
+    }
+
+    public static MBeanServer getMBeanServer() {
+        List<MBeanServer> mBeanServers = MBeanServerFactory.findMBeanServer(null);
+        if (!mBeanServers.isEmpty()) {
+            return mBeanServers.get(0);
+        }
+        return ManagementFactory.getPlatformMBeanServer();
+    }
+
+    @Override
+    public void bindTo(MeterRegistry reg) {
+        registerConsumerFetchMetrics(reg);
+        registerConsumerCoordinatorMetrics(reg);
+    }
+
+    private void registerConsumerFetchMetrics(MeterRegistry registry) {
+
+        registerMetricsEventually("type", "consumer-fetch-manager-metrics", (name, allTags) -> {
+
+                Gauge.builder("kafka.records.lag.max", mBeanServer,
+                    s -> safeDouble(() -> s.getAttribute(name, "records-lag-max")))
+                    .tags(allTags)
+                    .register(registry);
+
+                Gauge.builder("kafka.fetch.latency.avg", mBeanServer,
+                    s -> safeDouble(() -> s.getAttribute(name, "fetch-latency-avg")))
+                    .tags(allTags)
+                    .register(registry);
+
+                Gauge.builder("kafka.bytes.consumed.rate", mBeanServer,
+                    s -> safeDouble(() -> s.getAttribute(name, "bytes-consumed-rate")))
+                    .tags(allTags)
+                    .baseUnit("bytes")
+                    .register(registry);
+
+                Gauge.builder("kafka.fetch.size.max", mBeanServer,
+                    s -> safeDouble(() -> s.getAttribute(name, "fetch-size-max")))
+                    .tags(allTags)
+                    .baseUnit("bytes")
+                    .register(registry);
+
+                Gauge.builder("kafka.records.consumed.rate", mBeanServer,
```

Comment:
kafka publishes "rate" values as well as corresponding "total" values: 

https://kafka.apache.org/documentation/#monitoring

The rate captures the per/second  rate whereas the "total" is a running total (monotonically increasing). the problem with using the per second rates with a gauge is that you miss data, if you have a burst for a couple seconds that may not be reflected in the gauge stats. so i think in this case using the "total" values with a function counter actually makes more sense (this applies to all rate values below as well).

Reply:
Checking this further..

Reply:
Wherever it is possible to use a total rather than a rate coming from Kafka, we should do that. Micrometer already ships rates for monotonically increasing counters for those monitoring systems that don't provide query-side rate functions.